### PR TITLE
[MachinePipeliner] Make Recurrence MII More Accurate

### PR DESCRIPTION
**Pull request by mmarjieh**
Current RecMII calculation is bigger than it needs to be. The calculation was refined in this patch.

### CODE REVIEW

**Review by kasuga-fj — COMMENTED** [review]
Thanks (actually, I found the same problem recently and was trying to fix it, so thanks for the help). The algorithm looks good to me. I left a few minor comments, could you check them out?

**Comment by ytmukai** [comment]
Can you add some tests that will show that RecMII is now being calculated correctly? It can be created by testing the debug output as follows:
https://github.com/llvm/llvm-project/blob/4a12722110abb2ccb98173c82a7d7b96a5c098e0/llvm/test/CodeGen/Hexagon/swp-resmii.ll#L5-L7

File: llvm/include/llvm/CodeGen/MachinePipeliner.h [hunk 1]
```diff
@@ -339,24 +341,58 @@ class NodeSet {
...
+    // Calculate the latency of this node set.
+    // Example to demonstrate the calculation:
+    // Given: N0 -> N1 -> N2 -> N0
+    // Edges:
+    // (N0 -> N1, 3)
+    // (N0 -> N1, 5)
+    // (N1 -> N2, 2)
+    // (N2 -> N0, 1)
+    // The total latency which is a lower bound of the recurrence MII is the
+    // longest patch from N0 back to N0 given only the edges of this node set.
```
[the remark this answers is not in the document]
Comment (mmarjieh) [reply llvm/include/llvm/CodeGen/MachinePipeliner.h:355]: Yes, thanks

File: llvm/include/llvm/CodeGen/MachinePipeliner.h [hunk 2]
```diff
@@ -339,24 +341,58 @@ class NodeSet {
...
+        // potentially loop carried then a back-edge exists between the last
+        // node and the first node that isn't modeled in the DAG. Handle it
+        // manually by adding 1 to the distance of the last node.
+        if (PI.getSUnit() != FirstNode || PI.getKind() != SDep::Order ||
+            !DAG->isLoopCarriedDep(LastNode, PI, false))
+          continue;
+        SUnitToDistance[FirstNode] =
+            std::max(SUnitToDistance[FirstNode], SUnitToDistance[LastNode] + 1);
+      }
+    }
```
[the remark this answers is not in the document]
Comment (mmarjieh) [reply llvm/include/llvm/CodeGen/MachinePipeliner.h]: Done

**Comment by mmarjieh** [comment]
> Can you add some tests that will show that RecMII is now being calculated correctly? It can be created by testing the debug output as follows:
> 
> https://github.com/llvm/llvm-project/blob/4a12722110abb2ccb98173c82a7d7b96a5c098e0/llvm/test/CodeGen/Hexagon/swp-resmii.ll#L5-L7

Done in the new commit.
See the state of the test before this patch here:
https://godbolt.org/z/aT7so4nMc

File: llvm/lib/CodeGen/MachinePipeliner.cpp [hunk 3]
```diff
@@ -1680,13 +1680,19 @@ void SwingSchedulerDAG::Circuits::createAdjacencyStructure(
...
     for (auto &PI : SUnits[i].Preds) {
-      if (!SUnits[i].getInstr()->mayStore() ||
+      if (PI.getKind() != SDep::Order ||
           !DAG->isLoopCarriedDep(&SUnits[i], PI, false))
         continue;
-      if (PI.getKind() == SDep::Order && PI.getSUnit()->getInstr()->mayLoad()) {
+      if ((SUnits[i].getInstr()->mayLoad() &&
+           PI.getSUnit()->getInstr()->mayStore()) ||
+          (SUnits[i].getInstr()->mayStore() &&
+           PI.getSUnit()->getInstr()->mayLoad())) {
```
Comment (kasuga-fj) [inline llvm/lib/CodeGen/MachinePipeliner.cpp:1695]: Oh sorry, I made a mistake in the previous comment https://github.com/llvm/llvm-project/pull/105475#discussion_r1728130892.  I meant to say store-to-store, not store-to-load.  My intention was to remove conditions based on the type of instruction.  In this case, I think no conditional branching is needed here.

I'm not sure what your comment "Store after store was handled above" means.  If you mean output-dependencies, they are not relevant: Output-dependence is a write-after-write dependence for physical registers. 

Also, could you please fix the following comment?
https://github.com/llvm/llvm-project/blob/17726a8bdbb952b30957aa462b9b18d0c724464c/llvm/include/llvm/CodeGen/MachinePipeliner.h#L376

File: llvm/test/CodeGen/PowerPC/sms-recmii.ll [hunk 4]
```diff
@@ -0,0 +1,48 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 4
+; RUN: llc < %s -mtriple=powerpc64le-unknown-linux-gnu -verify-machineinstrs\
+; RUN:       -mcpu=pwr9 --ppc-enable-pipeliner --debug-only=pipeliner 2>&1 | FileCheck %s
+
```
Comment (ytmukai) [inline llvm/test/CodeGen/PowerPC/sms-recmii.ll:4]: Thanks for adding the test.
-debug-only requires a build with assertions, so please add the following:
```
; REQUIRES: asserts
```

File: llvm/test/CodeGen/PowerPC/sms-recmii.ll [hunk 5]
```diff
@@ -0,0 +1,48 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 4
```
Comment (ytmukai) [inline llvm/test/CodeGen/PowerPC/sms-recmii.ll:1]: This line is unnecessary.

File: llvm/lib/CodeGen/MachinePipeliner.cpp (hunk 3 above)
Comment (mmarjieh) [reply llvm/lib/CodeGen/MachinePipeliner.cpp:1695]: Hi @kasuga-fj,
I did what you requested locally.
I have a couple of failed lit tests.
Starting investigating this one:
`llvm/test/CodeGen/Hexagon/swp-resmii-1.ll`

Attaching both debug prints before this change and after:
[old_log.txt](https://github.com/user-attachments/files/16836615/old_log.txt)
[new_log.txt](https://github.com/user-attachments/files/16836616/new_log.txt)

The main difference is having a new backedge from one store to its preceding store.
This causes a new recurrence:
image
Which leads to invalid node order.
image

I am not sure what to do with this. Can you see if there is something wrong in the new log?

On a related note, can we do this change in another patch and keep the adjacency matrix as is here?

Comment (kasuga-fj) [reply llvm/lib/CodeGen/MachinePipeliner.cpp]: Thank you for your trying and investigating! 
I haven't investigated the details, but I think several things are happening that lead to the problem you encountered. The root cause, I guess, is that the newly added back-edges are not taken into account in the processes after finding circuits. I think it takes some effort to fix these problems, so I agree with you that this should be fixed by other patches.

**Review by kasuga-fj — APPROVED** [review]
LGTM. Thanks!

File: llvm/lib/CodeGen/MachinePipeliner.cpp (hunk 3 above)
Comment (mmarjieh) [reply llvm/lib/CodeGen/MachinePipeliner.cpp]: I agree.
Thanks for reviewing!

**Comment by mmarjieh** [comment]
@kasuga-fj @ytmukai 
Can someone submit on my behalf? I don't have write permissions.
Make sure to squash the commits and use the following commit message:

[MachinePipeliner] Make Recurrence MII More Accurate

Current RecMII calculation is bigger than it needs to be.
The calculation was refined in this patch.

**Comment by kasuga-fj** [comment]
Done

**Pull request merged**